src/torrent.cpp

```diff
@@ -5678,61 +5680,81 @@ namespace {
 
 		boost::scoped_ptr<std::vector<boost::uint8_t> > p(j->buffer.priorities);
 
-		if (j->ret == piece_manager::no_error)
+		if (m_file_priority != *p)
 		{
-			TORRENT_ASSERT(!j->error);
 			m_file_priority = *p;
 			update_piece_priorities();
 		}
-		else
-		{
-			// in this case, some file priorities failed to get set
-			TORRENT_ASSERT(j->error);
-			TORRENT_ASSERT(!m_share_mode);
 
-			if (alerts().should_post<file_error_alert>())
-				alerts().emplace_alert<file_error_alert>(j->error.ec
-					, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+		if (!j->error) return;
 
-			set_error(j->error.ec, j->error.file);
-			pause();
-		}
+		// in this case, some file priorities failed to get set
+		TORRENT_ASSERT(j->error);
+		TORRENT_ASSERT(!m_share_mode);
+
+		if (alerts().should_post<file_error_alert>())
+			alerts().emplace_alert<file_error_alert>(j->error.ec
+				, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+
+		set_error(j->error.ec, j->error.file);
+		pause();
 	}
 
 	void torrent::prioritize_files(std::vector<int> const& files)
 	{
 		INVARIANT_CHECK;
 
+		if (is_seed()) return;
+
 		// store file priorities until we get the metadata
 		if (!valid_metadata())
 		{
 			m_file_priority.assign(files.begin(), files.end());
 			return;
 		}
 
+		// the vector need to have exactly one element for every file
+		// in the torrent
+		TORRENT_ASSERT(int(files.size()) == m_torrent_file->num_files());
+
 		std::vector<boost::uint8_t> const new_priority = fix_priorities(files, m_torrent_file->files());
 
-		if (!m_torrent_file->num_pieces() || is_seed() || new_priority == m_file_priority) { return; }
+		if (new_priority == m_file_priority)
+		{
+			return;
+		}
 
 		// storage may be NULL during shutdown
-		if (boost::shared_ptr<piece_manager> const tmp = m_storage)
+		if (m_torrent_file->num_pieces() > 0 && m_storage)
 		{
 			inc_refcount("file_priority");
-			m_ses.disk_thread().async_set_file_priority(tmp.get(), new_priority,
-				boost::bind(&torrent::on_file_priority, shared_from_this(), _1));
+			m_ses.disk_thread().async_set_file_priority(m_storage.get()
+				, new_priority, boost::bind(&torrent::on_file_priority, shared_from_this(), _1));
 		}
 	}
 
 	void torrent::set_file_priority(int index, int prio)
 	{
-		if (index < 0 || (valid_metadata() && index >= m_torrent_file->num_files())) { return; }
+		INVARIANT_CHECK;
```

Comment:
This will result in multiple invariant checks, since `torrent::prioritize_files` has one.

Reply:
I don't think it's that big of deal. These are not included in release builds. But I also think the way to fix this would be to factor out the common parts of `prioritize_files` into a `prioritize_files_impl` which does not have the invariant checks. But if the invariant check would fail, knowing in more detail exactly which function it failed in would be helpful.

Reply:
after a second pass over this patch, I think the common code is really just the submission of the disk job, so I detached them again. I think refactoring of this should be done in master, once this has merged

Reply:
Fine.